File: src/main/java/net/katsstuff/danmakucore/impl/danmakuvariant/DanmakuVariantGeneric.java
/*
 * This class was created by <Katrix>. It's distributed as
 * part of the DanmakuCore Mod. Get the Source Code in github:
 * https://github.com/Katrix-/DanmakuCore
 *
 * DanmakuCore is Open Source and distributed under the
 * the DanmakuCore license: https://github.com/Katrix-/DanmakuCore/blob/master/LICENSE.md
 */
package net.katsstuff.danmakucore.impl.danmakuvariant;

import java.util.function.Supplier;

import net.katsstuff.danmakucore.data.MovementData;
import net.katsstuff.danmakucore.data.RotationData;
import net.katsstuff.danmakucore.data.ShotData;
import net.katsstuff.danmakucore.data.Vector3;
import net.katsstuff.danmakucore.entity.danmaku.DanmakuVariant;

public class DanmakuVariantGeneric extends DanmakuVariant {

	private ShotData shotData = null;
	private final MovementData movement;
	private final RotationData rotation = RotationData.none();

	private final Supplier<ShotData> shotDataSupplier;

	@SuppressWarnings("WeakerAccess")
	public DanmakuVariantGeneric(String name, Supplier<ShotData> shotData, MovementData movement) {
		super(name);
		shotDataSupplier = shotData;
		this.movement = movement;
	}

	public DanmakuVariantGeneric(String name, Supplier<ShotData> shotData, double speed) {
		this(name, shotData, new MovementData(speed, speed, speed, 0.0D, Vector3.GravityZero()));
	}

	@SuppressWarnings("WeakerAccess")
	public DanmakuVariantGeneric(MovementData movement, Supplier<ShotData> shotDataSupplier) {
		this.movement = movement;
		this.shotDataSupplier = shotDataSupplier;
	}

	@SuppressWarnings("unused")
	public DanmakuVariantGeneric(Supplier<ShotData> shotData, double speed) {
		this(new MovementData(speed, speed, speed, 0.0D, Vector3.GravityZero()), shotData);
	}

	@Override
	public ShotData getShotData() {
		if(shotData == null) shotData = shotDataSupplier.get();
		return shotData;
	}

	@Override
	public MovementData getMovementData() {
		return movement;
	}

	@Override
	public RotationData getRotationData() {
		return rotation;
	}
}

File: src/main/scala/net/katsstuff/teamnightclipse/danmakucore/impl/danmakuvariant/DanmakuVariantGeneric.scala
/*
 * Copyright (C) 2018  Katrix
 * This file is part of DanmakuCore.
 *
 * DanmakuCore is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * DanmakuCore is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with DanmakuCore.  If not, see <https://www.gnu.org/licenses/>.
 */
package net.katsstuff.teamnightclipse.danmakucore.impl.danmakuvariant

import java.util.function.Supplier

import net.katsstuff.teamnightclipse.danmakucore.danmaku.DanmakuVariant
import net.katsstuff.teamnightclipse.danmakucore.data.{MovementData, RotationData, ShotData}
import net.katsstuff.teamnightclipse.danmakucore.scalastuff.DanCoreImplicits._

class DanmakuVariantGeneric(
    name: String,
    shotDataSup: () => ShotData,
    movement: MovementData,
    rotation: RotationData = RotationData.none
) extends DanmakuVariant(name) {

  private lazy val shotData = shotDataSup()

  override def getShotData: ShotData         = shotData
  override def getMovementData: MovementData = movement
  override def getRotationData: RotationData = rotation
}
object DanmakuVariantGeneric {
  def apply(
      name: String,
      shotDataSup: () => ShotData,
      movement: MovementData,
      rotation: RotationData = RotationData.none
  ) = new DanmakuVariantGeneric(name, shotDataSup, movement, rotation)

  def withSpeed(name: String, shotDataSup: () => ShotData, speed: Double, rotation: RotationData = RotationData.none) =
    new DanmakuVariantGeneric(name, shotDataSup, MovementData.constant(speed), rotation)

  def create(name: String, shotDataSup: Supplier[ShotData], movement: MovementData, rotation: RotationData) =
    new DanmakuVariantGeneric(name, shotDataSup.asScala, movement, rotation)

  def create(name: String, shotDataSup: Supplier[ShotData], movement: MovementData) =
    new DanmakuVariantGeneric(name, shotDataSup.asScala, movement)

  def create(name: String, shotDataSup: Supplier[ShotData], speed: Double) =
    new DanmakuVariantGeneric(name, shotDataSup.asScala, MovementData.constant(speed))
}
